monolithic_integration_test/tests/integration_test.rs

```diff
@@ -1,207 +1,399 @@
-use futures::channel::oneshot::Sender;
 use janus::{
-    hpke::test_util::generate_hpke_config_and_private_key,
-    message::{Duration, Role, TaskId},
+    hpke::{
+        self, associated_data_for_aggregate_share, test_util::generate_hpke_config_and_private_key,
+        HpkeApplicationInfo, HpkePrivateKey, Label,
+    },
+    message::{Duration, HpkeConfig, Interval, Role, TaskId},
     task::VdafInstance,
     time::{Clock, RealClock},
+    TokioRuntime,
 };
 use janus_client::{self, Client, ClientParameters};
 use janus_server::{
-    aggregator::aggregator_server,
+    aggregator::{
+        aggregate_share::CollectJobDriver, aggregation_job_creator::AggregationJobCreator,
+        aggregation_job_driver::AggregationJobDriver, aggregator_server,
+    },
+    binary_utils::job_driver::JobDriver,
     datastore::{Crypter, Datastore},
+    message::{CollectReq, CollectResp},
     task::{test_util::generate_aggregator_auth_token, Task, PRIO3_AES128_VERIFY_KEY_LENGTH},
-    trace::{install_trace_subscriber, TraceConfiguration},
 };
-use prio::vdaf::prio3::{Prio3, Prio3Aes128Count};
+use janus_test_util::install_test_trace_subscriber;
+use prio::{
+    codec::{Decode, Encode},
+    field::Field64,
+    vdaf::{
+        prio3::{Prio3, Prio3Aes128Count},
+        AggregateShare, Collector,
+    },
+};
 use rand::{thread_rng, Rng};
+use reqwest::{
+    header::{self, CONTENT_TYPE},
+    redirect, StatusCode,
+};
 use std::{
     net::{IpAddr, Ipv4Addr, SocketAddr},
-    sync::Arc,
+    sync::{mpsc, Arc},
 };
-use tokio::task::JoinHandle;
+use tokio::{select, sync::oneshot, task, time, try_join};
 use url::Url;
 
 janus_test_util::define_ephemeral_datastore!();
 
-fn endpoint_from_socket_addr(addr: &SocketAddr) -> Url {
-    assert!(addr.ip().is_loopback());
-    let mut endpoint: Url = "http://localhost".parse().unwrap();
-    endpoint.set_port(Some(addr.port())).unwrap();
+#[tokio::test(flavor = "multi_thread")]
+async fn end_to_end() {
+    // Create a test case, and connect to the leader's database.
+    let test_case = TestCase::new().await;
+
+    // Upload a measurement, recording a timestamp before measurement upload to allow us to
+    // determine the correct collect interval.
+    let clock = RealClock::default();
+    const NUM_NONZERO_MEASUREMENTS: usize = 23;
+    let before_timestamp = clock.now();
+    for _ in 0..NUM_NONZERO_MEASUREMENTS {
+        test_case.client.upload(&0).await.unwrap();
+        test_case.client.upload(&1).await.unwrap();
+    }
+
+    // Wait long enough for aggregation jobs to be created.
+    time::sleep(time::Duration::from_secs(2)).await;
```

Comment:
First, the easy sleep: the second sleep isn't necessary, it is standing in as a replacement for polling. I went ahead and replaced it with a simple polling implementation.

...and then I noticed that the first sleep isn't necessary either. I believed it was because I saw a failure when I removed it, but I now suspect the failure was because I wasn't sleeping long enough for the overall aggregate+collect to complete successfully without it; with the polling implementation, we'll now wait for both operations.